### PR TITLE
Avoid re-emitting docgen for re-exported components

### PR DESCRIPTION
**Pull request by wrose504**
In our codebase we will often define a component in one file, then re-export it from `index.ts`. In some cases, the re-export uses a name that does not match the original component. In these cases, `react-docgen-loader` emits invalid code that tries to assign docgen info to the actual name from the original component, which fails.

    // components/Button/Button.tsx
    import React from 'react';
    export const Button = () => (<button type='submit'/>);
    export default Button;
    
    // components/Button/index.ts
    import ButtonComponent from './Button';
    export default ButtonComponent;

    // components/Bouton/index.ts
    import {Button as Bouton} from '../Button/Button';
    export {Bouton};
   
    // App.tsx
    import Button from './components/Button';
    import Bouton from './components/Bouton';
    
    const App = () => (
      <form><Button/><Bouton/></form>
    );

Currently invalid code is written into `components/Button/index.ts` and `components/Bouton/index.ts` that tries to assign to `Button.__docgenInfo`, when `Button` is not a valid identifier in either file.

The `react-webpack` implementation of `react-docgen-loader` avoids this problem by checking that the resolved component definition comes from the initial source file. See https://github.com/storybookjs/storybook/blob/e812f77def55aa7aeda70dddcc378c736412472d/code/presets/react-webpack/src/loaders/react-docgen-loader.ts#L25

This PR adds the same logic here, which fixes the issue.

### CODE REVIEW

File: sandboxes/react-rsbuild/src/stories/Bouton.stories.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import type { Meta, StoryObj } from '@storybook/react'
+import { fn } from '@storybook/test'
+import Bouton from './Bouton'
+
+// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
+const meta = {
+  title: 'Example/Bouton',
```
Comment (fi3ework) [inline sandboxes/react-rsbuild/src/stories/Bouton.stories.ts:7]: I think better to rename the new added story's title and file name to `ReExportButton` to help demonstrating its purpose.

**Review by fi3ework — CHANGES_REQUESTED** [review]

**Comment by fi3ework** [comment]
Thanks for the sync up with storybook's latest code. 🫶

File: sandboxes/react-rsbuild/src/stories/Bouton.ts [hunk 2]
```diff
@@ -0,0 +1,2 @@
+import { Button as Bouton } from './Button'
```
[the remark this answers is not in the document]
Comment (wrose504) [reply sandboxes/react-rsbuild/src/stories/Bouton.ts]: This is necessary to demonstrate the problem: if we just import `Button` named as it is, the generated code will not fail. I will update it to import as `ReExportButton` as you suggested.

**Review by fi3ework — APPROVED** [review]

**Comment by fi3ework** [comment]
Thanks you!

**Pull request merged**